### ISSUE TITLE
Incorrect copying in the FUSE module

### ISSUE THREAD

**Issue by arvgord** [billziss-gh/winfsp#254]
## Bug Report

Hi,

I use jnr-fuse for developing a network file system.  I implemented sending a file to the server in the `flush` method. It works, but I found that when I copy a file I see the (in cmd using copy or using Windows explorer) message "* files copied" **before** the `flush` method is done and zero(if an upload was successful) or error code(if an upload was failed) retrieved. 
In Linux, I get a message after the `flush` method completes.

### How to Reproduce

I tested it using WinFsp 2019.2 and the latest WinFsp version and it works similarly.

### Behaviors

Correct behavior of copy method. Return a message(copy result) **after** the flush method completes.

### Environment

- OS version and build: Windows 10x64 1809
- WinFsp version and build: WinFsp 2019.2 (also tested in WinFsp 2019.3 B2), WinFsp-FUSE version 2.8

**Comment by fsgeek** [comment]
Normally, copying files will be through the cache; the Windows Cache Manager then writes the (cached) pages back to the file system.  To disable this behavior, you would either need to disable write-back caching or use non-buffered I/O.  For example if you use `xcopy /j foo.dat bar.dat` it will use unbuffered I/O.  I would expect you would see the behavior you are expecting.

Linux (following POSIX) does not guarantee that writes are synchronous either (http://man7.org/linux/man-pages/man2/close.2.html) but the introduction of a **fsync** would certainly do so.  I spent a bit of time looking at the GNU implementation of cp and it is certainly not a simple "open a file/read a file/write a file" paradigm, which makes it more difficult to ascertain exactly what it is doing.

Bottom line, I don't think either behavior is "more correct", just "different".

**Comment by arvgord** [comment]
I use some options to prevent caching:
**attr_timeout=0
ac_attr_timeout=0
entry_timeout=0**
Also tried:
**direct_io
FileInfoTimeout=0
DirInfoTimeout=0
EaInfoTimeout=0
VolumeInfoTImeout=0**
But still, I get a message that a file was copied before the `flush` method is done.

**Comment by arvgord** [comment]
Also, I tried to use xcopy but get the same behavior.

**Comment by arvgord** [comment]
I got the same behavior using the [MemoryFS example](https://github.com/SerCeMan/jnr-fuse/blob/master/src/main/java/ru/serce/jnrfuse/examples/MemoryFS.java) in the jnr-fuse repository.

**Comment by billziss-gh** [comment]
@arvgord what you are experiencing is expected behavior and not a bug. Windows is not Linux and in general you should not expect behavioral parity between Linux FUSE and WinFsp FUSE at that level.

On Linux FUSE `flush` is sent to the file system prior to `release` and the file system has a chance to return an error code (if, for example, you cannot upload a file into a remote storage). This is to allow applications to check the return code of `close(2)` and detect a prior `write(2)` error. But notice that even Linux defers writes and it is perfectly possible for I/O calls to report success, but the I/O not make it to the storage medium.

Furthermore even on Linux FUSE it is possible to not receive `flush` after a `write` and prior to `release`. From https://sourceforge.net/p/fuse/mailman/message/32917430/:

> I'm asking because one of our internal systems (using the high-level
> API) would not get a flush in the following sequence:
>
>  open
>  mmap
>  close
>  (write)
>  munmap

On Windows there are two IRP's (I/O Request Packets) that may get sent when a file is closed. The first one is `IRP_MJ_CLEANUP` which for practical purposes is sent when `CloseHandle` happens. The second one is `IRP_MJ_CLOSE` which is sent when the last reference to the file object goes away; under normal circumstances this happens at an indeterminate time.

**One important thing to note is that neither `IRP_MJ_CLEANUP` nor `IRP_MJ_CLOSE` can report failure to the OS (i.e. the OS assumes that they always succeed).** This has the unfortunate consequence that the `flush` return value is necessarily ignored no matter how it is implemented.

On WinFsp-FUSE the `flush` happens at `IRP_MJ_CLOSE` time for reasons outside the scope of this discussion. Now this `IRP_MJ_CLOSE` is sent to the FSD when the OS decides to, which might well be after the process that issued the `CloseHandle` has long exited. Under WinFsp-FUSE (and if you do not specify `-o KeepFileCache`), the FSD asks the OS to send the `IRP_MJ_CLOSE` as soon as possible, but this may still not be soon enough. But recall that this is irrelevant anyway, because `IRP_MJ_CLOSE` cannot report failure and therefore the `flush` return code is just discarded!

----

You are probably thinking: alright, Bill, this is all well and good, but what do I do when I fail to upload my file to the remote storage? How do I report the error to the original application?

The answer is that you do not report any errors. Instead you design your file system in such a way as to be resilient against such problems. First you should handle remote uploads in `release` and not `flush` (recall that even on Linux there is no guarantee that `flush` comes before `release`). Second you should build a retry mechanism into your file system, that retries failed uploads periodically (even when the file system is restarted).

**Comment by billziss-gh** [comment]
@argvord I am wondering if the explanation above answers your questions.

**Comment by arvgord** [comment]
Thanks a lot, @billziss-gh! I think we have no way (workaround) to fix this in the WinWsp fuse module. I'm closing this issue.
